Introduction to Free Monads/article.md

```diff
@@ -0,0 +1,894 @@
+  pure ()
+```
+
+There is a catch, however.
+This code will happily compile, but it doesn't do anything interesting.
+Indeed, we didn't define the meaning of the computation, only its form, and we can't directly use `runStateF` to "run" our free state monad.
+
+This means we need to write an interpreter, so let's do that.
+Remember that `Free` has two constructors, `Pure` and `Free`, so we need to pattern-match on those:
+
+```haskell
+runState :: State s a -> s -> (a, s)
+runState (Pure x) s = (x, s)
+runState (Free f) s =
+  let (m, s') = runStateF f s
+  in runState m s'
+```
+
+If you're familiar with the usual state monad, you may have noticed that we've essentially moved the implementation of `>>=` to `runState`.
+The free monad doesn't specify the meaning of monadic actions, so we have to decide what those actions mean when we're running it.
+
+To illustrate this point, we're going to write another interpreter, which is essentially a pretty-printer for the flow of a `State` computation:
+
+```haskell
+printState :: (Show s, Show a) => State s a -> s -> String
+printState (Pure x) s = "pure (" <> show x <> "," <> show s <> ")"
+printState (Free m) s =
+  let (x, s') = runStateF m s
+  in "state change " <> show s <> " -> " <> show s' <> "\n"
+    <> printState x s'
+```
+
+If we run `someComputation` defined above through this interpreter, for example by invoking `printState someComputation 1`, we'll get the following output:
+
+```
+state change 1 -> 1
+state change 1 -> 2
+pure ((),2)
+```
+
+The first line of output corresponds to `get`.
+While it doesn't modify the state, it's still an action, so the `state change` line is still printed.
+The second line corresponds to `put $ i + 1`, which naturally changes the state to `2`.
+The third line corresponds to `pure ()`, but it also outputs the end state.
+
+Let's summarize what we've learned so far.
+We can take any usual base functor for some monad and get a `Monad` instance "for free".
+There's no free lunch here, though, so we need to define the semantics of this monad separately by writing an interpreter (or several if we need to).
+
+### List as a free monad
+
+Not all free monads behave exactly the same as their regular counterparts.
+For instance, you might know that the list monad encodes non-determinism[^non-determinism].
+However, if we derive a free monad for lists, the behavior is slightly different.
+Consider:
+
+[^non-determinism]: If the statement about list monad encoding non-determinism is surprising, consider that one could use a list to represent all possible outcomes of some event and then handle those one by one.
+    For further reading on this topic, refer to the [School of Haskell article on the list monad][school-of-haskell] or the [Haskell wikibook section on understanding monads][haskell-wikibook]
+
+    [school-of-haskell]: https://www.schoolofhaskell.com/school/starting-with-haskell/basics-of-haskell/13-the-list-monad#non-deterministic-computations
```

Comment:
Same thing as in  https://github.com/serokell/blog-posts/pull/8/files#r789621026